### PR TITLE
Use charges for determining valence

### PR DESCRIPTION
**Pull request by pckroon**
Here I figure out the valence of atoms by assigning electrons to orbital shells, and figuring out with how many unpaired electrons we have left.
Currently, I assume all paired electrons can be excited to higher shells, which deals with multivalent elements (S, P, ...). This does mean that N can also make 5 bonds if you ask it nicely. The *real* solution would be to figure out the balance between the excitation energy on one hand, and the electron unpairing + bond energy on the other, but that sounds really hard. So either we accept this slightly strange behaviour, or we limit this excitation to higher shells only. Further consideration: should the excitation be limited to just 2 electrons, to the next s-orbital?

Still a draft, since I need to iron out some failing tests.

### CODE REVIEW

**Comment by fgrunewald** [comment]
Things I feel like should raise Errors:

- [NH5]
- O=S(=O)(=O)=O
- O=S(=O)(=O)(=O)=O

**Comment by pckroon** [comment]
I do like the idea of breaking on overvalence atoms, but I think that should be a separate PR

File: pysmiles/smiles_helper.py [hunk 1]
```diff
@@ -342,37 +347,70 @@ def bonds_missing(mol, node_idx, use_order=True):
...
+    return int(max(val[0] - bonds, 0))
 
 
-def _valence(mol, node_idx, minimum=0):
+def valence(atom):
     """
-    Returns the valence of the specified node. Since some elements can have
-    multiple valences, give the smallest one that is more than `minimum`.
+    Returns the valence of the atom. Since some elements can have
+    multiple valences, The valence is returned as list.
```
Comment (fgrunewald) [inline pysmiles/smiles_helper.py:359]: Typo replace `,` by `.` in 359

File: pysmiles/smiles_helper.py [hunk 2]
```diff
@@ -342,37 +347,70 @@ def bonds_missing(mol, node_idx, use_order=True):
...
+    # Any electrons we have leftover we distribute over the orbitals. First 1
+    # electron in each, then we start making pairs. The resulting valence will
+    # be the number of unpaired electrons. Added bonus/complication: electrons
+    # in pairs we can excite to higher shells, increasing the number of unpaired
+    # electrons
+    shell = ORBITAL_SIZES[idx]
+    shell_size = sum(shell)
+    single_electrons, paired_electrons = divmod(electrons, shell_size//2)
+    if single_electrons == 0:
+        single_electrons, paired_electrons, empty_orbitals = paired_electrons, 0, shell_size//2 - paired_electrons
```
Comment (fgrunewald) [inline pysmiles/smiles_helper.py:399]: what is the purpose of empty_orbitals here? why not say `single_electrons = shell_size//2` and `paired_electrons = 0`

**Review by fgrunewald — CHANGES_REQUESTED** [review]
I like the algorithm a lot! This should solve most of our problems. Only comment is that the orbital counting code is super confusing when you need to understand it the first time around. 

if you read the periodic table you could also get rid of the `organic_subset` variable in the read_smiles function. It's not needed but could be a nice addition.

File: pysmiles/smiles_helper.py [hunk 3]
```diff
@@ -342,37 +347,70 @@ def bonds_missing(mol, node_idx, use_order=True):
...
+    elif single_electrons == 1:
+        single_electrons = shell_size//2 - paired_electrons
+        single_electrons, paired_electrons, empty_orbitals = single_electrons, paired_electrons, 0
+    else:  # pragma: nocover
+        raise AssertionError(f"{single_electrons=}")
```
[the remark this answers is not in the document]
Comment (pckroon) [reply pysmiles/smiles_helper.py:404]: No. If that ever breaks (wouldn't know how) it'll print the name of the variable with the value

File: pysmiles/smiles_helper.py (hunk 2 above)
Comment (pckroon) [reply pysmiles/smiles_helper.py:399]: It's mostly for later (YAGNI!), if we want to be able to adjust the charge to fix the valence of atoms.

**Review by fgrunewald — APPROVED** [review]
looks good to me

**Pull request merged**